controllers/vaultsecret_controller.go

```diff
@@ -0,0 +1,307 @@
+		if !ok {
+			return nil
+		}
+		// if the resync period has not elapsed, do not reconcile
+		if !instance.Status.LastVaultSecretUpdate.Add(duration).Before(time.Now()) {
+			return nil
+		}
+	}
+
+	r.Log.V(1).Info("Reconcile VaultSecret", "namespacedName", fmt.Sprintf("%v/%v", instance.Namespace, instance.Name))
+
+	mergedMap := make(map[string]interface{})
+
+	definitionsStatus := make([]redhatcopv1alpha1.VaultSecretDefinitionStatus, len(instance.Spec.VaultSecretDefinitions))
+
+	for idx, vaultSecretDefinition := range instance.Spec.VaultSecretDefinitions {
+		vaultClient, err := vaultSecretDefinition.Authentication.GetVaultClient(ctx, instance.Namespace)
+		if err != nil {
+			r.Log.Error(err, "unable to create vault client", "instance", instance)
+			return err
+		}
+
+		ctx = context.WithValue(ctx, "vaultClient", vaultClient)
+		vaultEndpoint := vaultutils.NewVaultEndpointObj(&vaultSecretDefinition)
+		vaultSecret, ok, _ := vaultEndpoint.ReadSecret(ctx)
+		if !ok {
+			return errors.New("unable to read Vault Secret for " + vaultSecretDefinition.GetPath())
+		}
+
+		definitionsStatus[idx] = redhatcopv1alpha1.VaultSecretDefinitionStatus{
+			Name:          vaultSecretDefinition.Name,
+			LeaseID:       vaultSecret.LeaseID,
+			LeaseDuration: vaultSecret.LeaseDuration,
+			Renewable:     vaultSecret.Renewable,
+		}
+
+		mergedMap[vaultSecretDefinition.Name] = vaultSecret.Data
+	}
+
+	k8sSecret, err := r.formatK8sSecret(instance, mergedMap)
+	if err != nil {
+		r.Log.Error(err, "unable to format k8s secret", "instance", instance)
+		return err
+	}
+
+	err = r.CreateOrUpdateResource(ctx, instance, instance.GetNamespace(), k8sSecret)
+	if err != nil {
+		return err
+	}
+
+	now := metav1.NewTime(time.Now())
+	instance.Status.LastVaultSecretUpdate = &now
+	instance.Status.VaultSecretDefinitionsStatus = definitionsStatus
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *VaultSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	vaultSecretPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			newVaultSecret, ok := e.ObjectNew.DeepCopyObject().(*redhatcopv1alpha1.VaultSecret)
+			if !ok {
+				return false
+			}
+			oldVaultSecret, ok := e.ObjectOld.DeepCopyObject().(*redhatcopv1alpha1.VaultSecret)
+			if !ok {
+				return false
+			}
+
+			if !reflect.DeepEqual(oldVaultSecret.Spec, newVaultSecret.Spec) {
+				r.Log.V(1).Info("Update Event - Spec changed", "object", e.ObjectNew)
+				return true
+			}
+			return false
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			r.Log.V(1).Info("Create Event", "object", e.Object)
+			return true
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			r.Log.V(1).Info("Delete Event", "object", e.Object)
+			return true
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	k8sSecretPredicate := predicate.Funcs{
```

Comment:
I don't understand this, why are you adding the `Owns` clause and then discarding all of the events?

Reply:
yeah I probably don't need the Owns anymore. ill test removing it.